File: spec/test_token_user_fields.c
#include "testparser.h"
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int main()
{
    char const * input =
        "# c1\n"
        "#  c2\n"
        "\n"
        "first\n"
        "\n   \n  \n"
        "  # s1\n"
        "   #   s2\n"
        "second\n";
    p_context_t * context;
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_SUCCESS);
    Start start = p_result(context);
    IDs ids = p_Start_pIDs(start);
    assert(p_node_valid(ids));
    Token id0 = p_IDs_id(ids);
    assert(p_node_valid(id0));
    assert(p_node_data(id0)->comments);
    assert(strcmp(p_node_data(id0)->comments, "# c1\n#  c2\n") == 0);
    IDs ids2 = p_IDs_pIDs(ids);
    assert(p_node_valid(ids2));
    Token id1 = p_IDs_id(ids2);
    assert(p_node_valid(id1));
    assert(p_node_data(id1)->comments);
    assert(strcmp(p_node_data(id1)->comments, "# s1\n#   s2\n") == 0);

    free(context->comments);
    p_context_delete(context);

    return 0;
}

File: spec/test_token_user_fields.rs
use testparser::*;

fn main() {
    let input = b"# c1\n#  c2\n\nfirst\n\n   \n  \n  # s1\n   #   s2\nsecond\n";
    let mut c = p_context_new(input);
    assert_eq!(P_SUCCESS, p_parse(&mut c));
    {
        let start = p_result(&c);
        assert!(start.pIDs().valid());
        assert!(start.pIDs().id().valid());
        assert_eq!("# c1\n#  c2\n", start.pIDs().id().data().comments.as_str());
        assert!(start.pIDs().pIDs().valid());
        assert!(start.pIDs().pIDs().id().valid());
        assert_eq!("# s1\n#   s2\n", start.pIDs().pIDs().id().data().comments.as_str());
    }
    p_context_delete(c);
}
